compressors/lz77.py

```diff
@@ -0,0 +1,705 @@
+            + combined_encoding
+        )
+
+    def encode_literals(self, literals: List):
+        """Perform entropy encoding of the literals and return the encoded bitarray.
+
+        We apply Huffman coding for the literals and also store the counts to enable the
+        decoder to construct the same Huffman tree.
+
+        Args:
+            encoded_bitarray (BitArray): encoded bit array
+        """
+        # first encode the literals with empirical Huffman code
+        counts = DataBlock(literals).get_counts()
+        if len(counts) > 0:
+            prob_dict = ProbabilityDist.normalize_prob_dict(counts).prob_dict
+            # let's sort prob_dist by the alphabet to make sure we get exact same ordering
+            # during decompression as well, which makes sure it's the same Huffman tree!
+            prob_dist_sorted = ProbabilityDist({i: prob_dict[i] for i in sorted(prob_dict)})
+
+            literals_encoding = HuffmanEncoder(prob_dist_sorted).encode_block(DataBlock(literals))
+
+            # Now encode the counts using Elias Delta code.
+            # We first generate a list of counts (of length 256) and then apply Elias Delta.
+            # For any bytes not seen in the data, set count to 0.
+            for i in range(256):
+                if i not in counts:
+                    counts[i] = 0
+            counts_list = [counts[i] for i in range(256)]
+
+            counts_encoding = EliasDeltaUintEncoder().encode_block(DataBlock(counts_list))
+            # combine everything into a single bitarray
+            return (
+                uint_to_bitarray(len(counts_encoding), ENCODED_BLOCK_SIZE_HEADER_BITS)
+                + counts_encoding
+                + uint_to_bitarray(len(literals_encoding), ENCODED_BLOCK_SIZE_HEADER_BITS)
+                + literals_encoding
+            )
+        else:
+            # if no counts (i.e., no literals) just transmit 0
+            return uint_to_bitarray(0, ENCODED_BLOCK_SIZE_HEADER_BITS)
+
+    def find_match_length(self, start_pos_1: int, start_pos_2: int):
+        """Find the match length of window starting from start_pos_1 and start_pos_2.
+           That is, largest match_length s.t.
+           window[start_pos_1:start_pos_1+match_length]==window[start_pos_2:start_pos_2+match_length]
+           Note that matching sections are allowed to overlap
+
+        Args:
+            start_pos_1 (int)
+            start_pos_2 (int)
+
+        Returns:
+            int: match length
+        """
+        match_length = 0
+        while start_pos_1 + match_length < len(self.window) and start_pos_2 + match_length < len(
+            self.window
+        ):
+            if self.window[start_pos_1 + match_length] != self.window[start_pos_2 + match_length]:
+                break
+            else:
+                match_length += 1
+        return match_length
+
+    def lz77_parse_generate_sequences(self, data_block: DataBlock):
```

Comment:
minor -> should this be named `lz77_parse_and_generate_sequences`? (as currently it seems like you are parsing the generated sequences)

Reply:
done now